akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerProxy.scala

```diff
@@ -0,0 +1,258 @@
+   * @param failureDetector      function to detect if the a message received from the target actor as
+   *                             response from a request represent a failure
+   * @param failureMap           function to map a failure into a response message. The failing response message is wrapped
+   *                             into a [[akka.contrib.circuitbreaker.CircuitBreakerProxy.CircuitOpenFailure]] object
+   */
+  def props(target: ActorRef,
+            maxFailures: Int,
+            callTimeout: Timeout,
+            resetTimeout: Timeout,
+            circuitEventListener: Option[ActorRef],
+            failureDetector: Any ⇒ Boolean,
+            failureMap: CircuitOpenFailure ⇒ Any) =
+    Props(new CircuitBreakerProxy(target, maxFailures, callTimeout, resetTimeout, circuitEventListener, failureDetector, failureMap))
+
+  sealed trait CircuitBreakerCommand
+
+  final case class TellOnly(msg: Any) extends CircuitBreakerCommand
+  final case class Passthrough(msg: Any) extends CircuitBreakerCommand
+
+  sealed trait CircuitBreakerResponse
+  final case class CircuitOpenFailure(failedMsg: Any)
+
+  sealed trait CircuitBreakerEvent
+  final case class CircuitOpen(circuit: ActorRef) extends CircuitBreakerCommand
+  final case class CircuitClosed(circuit: ActorRef) extends CircuitBreakerCommand
+  final case class CircuitHalfOpen(circuit: ActorRef) extends CircuitBreakerCommand
+
+  sealed trait CircuitBreakerState
+  case object Open extends CircuitBreakerState
+  case object Closed extends CircuitBreakerState
+  case object HalfOpen extends CircuitBreakerState
+
+  final case class CircuitBreakerStateData(failureCount: Int = 0, firstHalfOpenMessageSent: Boolean = false)
+
+  final case class CircuitBreakerPropsBuilder(
+    maxFailures: Int, callTimeout: Timeout, resetTimeout: Timeout,
+    circuitEventListener: Option[ActorRef] = None,
+    failureDetector: Any ⇒ Boolean = { _ ⇒ false },
+    openCircuitFailureConverter: CircuitOpenFailure ⇒ Any = identity) {
+
+    def withMaxFailures(value: Int) = copy(maxFailures = value)
+    def withCallTimeout(value: Timeout) = copy(callTimeout = value)
+    def withResetTimeout(value: Timeout) = copy(resetTimeout = value)
+    def withCircuitEventListener(value: Option[ActorRef]) = copy(circuitEventListener = value)
+    def withFailureDetector(value: Any ⇒ Boolean) = copy(failureDetector = value)
+    def withOpenCircuitFailureConverter(value: CircuitOpenFailure ⇒ Any) = copy(openCircuitFailureConverter = value)
+
+    /**
+     * Creates the props for a [[akka.contrib.circuitbreaker.CircuitBreakerProxy]] proxying the given target
+     *
+     * @param target the target actor ref
+     */
+    def props(target: ActorRef) = CircuitBreakerProxy.props(target, maxFailures, callTimeout, resetTimeout, circuitEventListener, failureDetector, openCircuitFailureConverter)
+
+  }
+}
+
+object CircuitBreakerInternalEvents {
```

Comment:
Private to CircuitBreakerProxy inside the companion object, since they are internal?